### ISSUE TITLE
Python/ABI tag wrong on all platforms

### ISSUE THREAD

**Issue by peter-urban** [FFY00/meson-python#95]
The platform tag for my module is set to '-py-none-' on all platforms despite creating a shared module.
Apart from the tag issue the installation goes well and the share module is included in the package
This only happens if no other shared or platlib libraries are included

I can trace the issue as follows (on linux):
in __init__.py
self._install_plan.get('targets', {}) does read the full installation path rather than {moduledir_shared} as destination of my shared module
This is weird for me, because when I run pip wheel --cache-dir 'pybuilddir' and manually read the "intro-install_plan.json" file in the pybuilddir, I see {moduledir_shared} in front of the shared module

The problem then happens in function _files_by_tag (_init__.py line 727): 
```
if (
    not mesonpy._compat.is_relative_to(destination, '{py_platlib}')
    and not mesonpy._compat.is_relative_to(destination, '{moduledir_shared}')
):
    continue
tag = self._calculate_file_abi_tag_heuristic(file)
```
Because the destination is not relative to '{moduledir_shared}' it never reaches the point _calculate_file_abi_tag_heuristic(file) for the module file. 
Because of this:
- the function  _select_abi_tag [line 737] receives no tags 
- and thus the functions python_tag [line 647] and abi_tag [line 654] return 'py' and 'none'

Any idea why my shared module destination is not detected as {moduledir_shared}?

**Comment by rgommers** [comment]
I see this too with SciPy. Things work as they should with Meson `0.62.2`, and don't with Meson `0.63.0` (released 3 days ago). 

If I use Meson `0.63.0` directly, it still produces `.cpython-310-x86_64-linux-gnu.so` extensions.

Aside from the fix, it would be great to do some more consistency checks and start erroring out in case of issues. If `is_pure` is false (already known), then `abi_tag()` should never return `none`. This metadata in the `WHEEL` file can never be correct I believe:
```
Root-Is-Purelib: false
Tag: py3-none-linux_x86_64
```

**Comment by peter-urban** [comment]
> Things work as they should with Meson 0.62.2, and don't with Meson 0.63.0 (released 3 days ago).
Thx for clarifying this! I was wondering why my CI chain stopped working. Now I realize that it automatically upgraded to the latest meson.

**Comment by rgommers** [comment]
@FFY00 explained that the "pure or not" determination inside `meson-python` is based on install directory metadata returned from `meson`. The regression may be because Meson stopped using the `{moduledir_shared}`-placeholder. 

We also chatted about pure vs. Python-specific builds in general. Right now the detection for that is fragile; for Meson this is to some extent unavoidable since it doesn't know it is building a single package (it could be building multiple independent targets and standalone scripts that get installed outside of `site-packages` as, e.g. command-line tools. However, `meson-python` knows it is building a single package, so perhaps simply defaulting to Python-specific builds and have a single boolean setting for pure wheels in `pyproject.toml` makes more sense than trying to infer this information.

**Comment by eli-schwartz** [comment]
> because Meson stopped using the `{moduledir_shared}`-placeholder

It should still be doing this? The only change to this logic was to avoid cases where something is installed to `{moduledir_shared}/foo/bar.so` and the install_plan reported it at `{moduledir_shared}/bar.so`

According to https://github.com/mesonbuild/meson/pull/9474 this was causing problems... I reworked the logic so that in cases where it misdetected the install directory it would expand an accurate location, which according to my understanding was supposed to be a partial fix.

I'll confess that I'm not actually 100% sure how this is being used... I think the idea was that this build backend was reading the install plan and installing files itself based on the placeholder, and heuristically guessing at the rest?

**Comment by FFY00** [comment]
That is a good change. I will look into this to figure out what is the root cause.

**Comment by FFY00** [comment]
It seems the issue is that now we are getting the full path instead of one with a placeholder.

```json
{
  "targets": {
    "/home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/builddir/plat.cpython-310-x86_64-linux-gnu.so": {
      "destination": "/usr/lib/python3.10/site-packages/plat.cpython-310-x86_64-linux-gnu.so",
      "tag": "runtime"
    }
  },
  "python": {
    "/home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/pure.py": {
      "destination": "{py_purelib}/pure.py",
      "tag": "runtime"
    }
  }
}
```

Previously
```json
{
  "targets": {
    "/home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/builddir/plat.cpython-310-x86_64-linux-gnu.so": {
      "destination": "{moduledir_shared}/plat.cpython-310-x86_64-linux-gnu.so",
      "tag": "runtime"
    }
  },
  "python": {
    "/home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/pure.py": {
      "destination": "{py_purelib}/pure.py",
      "tag": "runtime"
    }
  }
}
```

**Comment by FFY00** [comment]
This should have be fixed by #103

**Comment by rgommers** [comment]
This is not fixed in `meson-python` 0.7.0, still seeing the problem on Linux and macOS. The reproducer is straightforward with SciPy `main`:
```bash
$ mamba env create -f environment.yml  # in root of scipy repo
$ mamba activate scipy-dev
$ mamba install build
$ python -m build
...
Successfully built scipy-1.10.0.dev0.tar.gz and scipy-1.10.0.dev0-py3-none-linux_x86_64.whl
```

Then do `git clean -xdf`, apply this patch:
```diff
diff --git a/pyproject.toml b/pyproject.toml
index 44cde80ae..65c899267 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -11,6 +11,7 @@
 build-backend = 'mesonpy'
 requires = [
     "meson-python>=0.5.0",
+    "meson==0.62.2",
     "Cython>=0.29.21",
     "pybind11>=2.4.3",
     "pythran>=0.11.0",
```
and rebuild to see that things do work with meson `0.62.2`:
```
$ python -m build
...
Successfully built scipy-1.10.0.dev0.tar.gz and scipy-1.10.0.dev0-cp310-cp310-linux_x86_64.whl
```

**Comment by rgommers** [comment]
> However, `meson-python` knows it is building a single package, so perhaps simply defaulting to Python-specific builds and have a single boolean setting for pure wheels in `pyproject.toml` makes more sense than trying to infer this information.

This is also still relevant. Would you prefer a separate issue to keep track of this @FFY00?

**Comment by rgommers** [comment]
I am pinning `meson` to 0.62.2` for the SciPy 1.9.0 release to unblock that release, see https://github.com/scipy/scipy/pull/16671

**Comment by peter-urban** [comment]
> This is not fixed in meson-python 0.7.0, still seeing the problem on Linux and macOS. 

I just want to confirm that the changes in 0.7 did not resolve the issue for my project either

**Comment by FFY00** [comment]
I tracked this down to a super weird issue where `sysconfig.get_path('platlib')` is returning different paths throughout the life of the code :exploding_head:

**Comment by FFY00** [comment]
Okay, that is due to

https://github.com/FFY00/meson-python/blob/main/mesonpy/__init__.py#L236-L237

Apparently `sysconfig.get_config_vars()` returns the original dictionary.

**Comment by rgommers** [comment]
> This is not fixed in `meson-python` 0.7.0, still seeing the problem on Linux and macOS. The reproducer is straightforward with SciPy `main`:

I've tried this several times on both Linux and macOS to ensure I didn't do something obviously wrong, but: I can reproduce this _only_ when using `python -m build` (which picks up meson 0.63.0 and meson-python 0.7.0 from PyPI in a fresh virtualenv) and _not_ when I install those exact same versions in my conda env  (either from conda-forge or from PyPI, same result) and use `python -m build --no-isolation --skip-dependency-check`. So: this is either a path-related issue or some virtualenv-related weirdness with `sysconfig`.

**Comment by FFY00** [comment]
I reproduced this by creating a venv inside the conda environment and installing all the requirements. That is basically what pypa/build does anyway.

**Comment by rgommers** [comment]
Ah okay, that helps if the problem is due to being in a venv, rather that isolated vs. non-isolated.

**Comment by FFY00** [comment]
Well, an isolated build is just a build inside a virtual environment with just the dependencies :stuck_out_tongue:
